### PR TITLE
fix validate_path_is_safe function on windows

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

Update validate_path_is_safe function to work properly on windows machine.

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/test_rest_tracking.py [hunk 1]
```diff
@@ -544,6 +559,40 @@ def test_validate_path_is_safe_bad(path):
...
+        "C:\\path\\..\\to\\file",
+        "C:/path/../to/file"
+        # UNC(Universal Naming Convention) paths
+        "\\\\path\\to\\file",
+        "\\\\path/to/file",
+        "\\\\.\\C:\\path\\to\\file",
+        "\\\\?\\C:\\path\\to\\file",
+        "\\\\?\\UNC/path/to/file",
+    ],
+)
```
Comment (harupy) [inline tests/tracking/test_rest_tracking.py:590]: /ai rewrite this using raw string and chage `\\` to `\`

File: tests/tracking/test_rest_tracking.py [hunk 2]
```diff
@@ -528,6 +528,21 @@ def test_validate_path_is_safe_good(path):
...
 
 
+@pytest.mark.skipif(not is_windows(), reason="This test only passes on Windows")
+@pytest.mark.parametrize(
+    "path",
+    [
+        # relative path from current directory of C: drive
+        "C:path",
+        "C:path/",
+        "C:path/to/file",
```
Comment (harupy) [inline tests/tracking/test_rest_tracking.py:538]: Is `C:../path/to/file"` bad?

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:538]: emmm I don't think that's a valid path? 
image

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:538]: But if providing this path, it will be identified as a relative path 🤔

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:538]: good relative path or bad relative path?

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:538]: good relative path I think, it doesn't meet any of the three conditions to throw an exception

File: tests/tracking/test_rest_tracking.py [hunk 3]
```diff
@@ -544,6 +559,40 @@ def test_validate_path_is_safe_bad(path):
         validate_path_is_safe(path)
 
 
+@pytest.mark.skipif(not is_windows(), reason="This test only passes on Windows")
+@pytest.mark.parametrize(
+    "path",
+    [
+        "../path",
```
Comment (BenWilson2) [inline tests/tracking/test_rest_tracking.py:566]: some additional Windows-specific attack vectors:

```

    r'../etc/passwd',
    r'..\Windows\System32\',
    r'C:\Windows\System32\',
    r'/etc/passwd',
    r'::Windows\System32',
    r'..\..\..\..\Windows\System32\',
    r'../Windows/System32',
    r'.../...//',
    r'....\\',
    r'\\?\C:\Windows\System32\',
    r'\\.\C:\Windows\System32\',
    r'\\UNC\Server\Share\',
    r'\\Server\Share\folder\',
    r'\\127.0.0.1\c$\Windows\',
    r'\\localhost\c$\Windows\',
    r'\\smbserver\share\path\',
    r'..\\?\C:\Windows\System32\',
    r'file:///C:/Windows/System32/',
    r'C:/Windows/../Windows/System32/',
    r'C:\Windows\..\Windows\System32\',
    r'/Windows/no/such/path',
    r'../../../../../../../../../../../../Windows/System32',
    r'../../../../../../../../../../../../etc/passwd',
    r'../../../../../../../../../../../../var/www/html/index.html',
    r'../../../../../../../../../../../../usr/local/etc/openvpn/server.conf',
    r'../../../../../../../../../../../../Program Files (x86)',
    r'/../../../../../../../../../../../../Windows/System32',
    r'file:///etc/passwd',
    r'file:///d:/windows/repair/sam',
    r'file:///proc/version',
    r'file:///inetpub/wwwroot/global.asa',
    r'file://c:/boot.ini',
    r'/file://etc/passwd',
    r'/file:c:/passwd',
    r'/file://d:/windows/win.ini',
    r'/file://./windows/win.ini',
    r'/file://../windows/win.ini',
    r'/Windows\../etc/passwd',
    r'/Windows\..\Windows\System32\',
    r'/etc/passwd%00.jpg',
    r'/etc/passwd%00.html',
    r'/etc/passwd%00.txt',
    r'/etc/passwd%00.php',
    r'/etc/passwd%00.asp',
    r'/Windows\..\Windows\System32\cmd.exe',
    r'/Windows\..\Windows\System32\msconfig.exe',
    r'/Windows\..\Windows\System32\regedit.exe',
    r'/Windows\..\Windows\System32\taskmgr.exe',
    r'/Windows\..\Windows\System32\control.exe',
    r'/Windows\..\Windows\System32\services.msc',
    r'/Windows\..\Windows\System32\diskmgmt.msc',
    r'/Windows\..\Windows\System32\eventvwr.msc',
```

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:566]: image

I thought paths beginning with /xxx is always considered as a relative path on windows right? So it should be fine on windows? like "/path", it should be a good path.

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:566]: We can't believe what ChatGPT says.

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:566]: I'm just curious why /etc/passwd is not a good path in windows if it's a relative path?

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:566]: If it's a relative path, I think it's a good path.

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:566]: On windows, is it possible to join `/etc/passwd` and another path?

Comment (serena-ruan) [reply tests/tracking/test_rest_tracking.py:566]: I think so, path joining is regardless of relative or absolute path?

File: mlflow/server/handlers.py [hunk 4]
```diff
@@ -530,7 +530,7 @@ def validate_path_is_safe(path):
     if (
         any((s in path) for s in _OS_ALT_SEPS)
         or ".." in path.split(posixpath.sep)
-        or posixpath.isabs(path)
+        or pathlib.Path(path).is_absolute()
```
Comment (harupy) [inline mlflow/server/handlers.py:533]: btw do we need to support windows paths?

Comment (harupy) [reply mlflow/server/handlers.py:533]: or is the challenge posix path detection?

**Comment by serena-ruan** [comment]
Tests on windows machine:
Good paths:
image

File: mlflow/server/handlers.py (hunk 4 above)
Comment (serena-ruan) [reply mlflow/server/handlers.py:534]: Yes, on windows machine 
image

**Comment by adeadfed** [comment]
> Tests on windows machine: Good paths: image

Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`

While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:

 ```
 >>> import pathlib
>>> pathlib.Path('/test').is_absolute()
False
>>> pathlib.Path('/').is_absolute()
False
>>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
False
>>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
```

**Comment by adeadfed** [comment]
> > > Tests on windows machine: Good paths: image
> > 
> > 
> > Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`
> > While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:
> > ```
> > >>> import pathlib
> > >>> pathlib.Path('/test').is_absolute()
> > False
> > >>> pathlib.Path('/').is_absolute()
> > False
> > >>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
> > False
> > >>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
> > ['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
> > ```
> 
> Hi @adeadfed, thanks so much for this helpful information! Since you're an expert on this, would you be willing to add some test cases in your mind to the test suites so that we could better validate our approach? 😃

At quick glance, the list of paths in your tests is exhaustive imo. The inconsistent behavior of pathlib should be resolved like this:

```
def validate_path_is_safe(path):
    """
    Validates that the specified path is safe to join with a trusted prefix. This is a security
    measure to prevent path traversal attacks.
    """
    if is_file_uri(path):
        path = local_file_uri_to_path(path)
    if (
        any((s in path) for s in _OS_ALT_SEPS)
        or ".." in path.split(os.sep)
        or pathlib.PureWindowsPath(path).is_absolute()
        or pathlib.PurePosixPath(path).is_absolute()
    ):
        raise MlflowException(f"Invalid path: {path}", error_code=INVALID_PARAMETER_VALUE)
```

Though I am wondering if `or ".." in path.split(os.sep)` check can be bypassed now or not. I guess we should see that from the output of the tests :) Hope that helps!

**Comment by serena-ruan** [comment]
> > > > Tests on windows machine: Good paths: image
> > > 
> > > 
> > > Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`
> > > While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:
> > > ```
> > > >>> import pathlib
> > > >>> pathlib.Path('/test').is_absolute()
> > > False
> > > >>> pathlib.Path('/').is_absolute()
> > > False
> > > >>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
> > > False
> > > >>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
> > > ['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
> > > ```
> > 
> > 
> > Hi @adeadfed, thanks so much for this helpful information! Since you're an expert on this, would you be willing to add some test cases in your mind to the test suites so that we could better validate our approach? 😃
> 
> At quick glance, the list of paths in your tests is exhaustive imo. The inconsistent behavior of pathlib should be resolved like this:
> 
> ```
> def validate_path_is_safe(path):
>     """
>     Validates that the specified path is safe to join with a trusted prefix. This is a security
>     measure to prevent path traversal attacks.
>     """
>     if is_file_uri(path):
>         path = local_file_uri_to_path(path)
>     if (
>         any((s in path) for s in _OS_ALT_SEPS)
>         or ".." in path.split(os.sep)
>         or pathlib.PureWindowsPath(path).is_absolute()
>         or pathlib.PurePosixPath(path).is_absolute()
>     ):
>         raise MlflowException(f"Invalid path: {path}", error_code=INVALID_PARAMETER_VALUE)
> ```
> 
> Though I am wondering if `or ".." in path.split(os.sep)` check can be bypassed now or not. I guess we should see that from the output of the tests :) Hope that helps!

Thanks! Updated and let's see how the tests run :)

**Comment by adeadfed** [comment]
Well, judging by the test output, pathlib now works fine. All failed test cases under the `test_validate_path_is_safe_windows_good` are false positives and will be indeed treated by Python as absolute paths, so they should better be moved to the `test_validate_path_is_safe_windows_bad` list of paths. However, there is indeed a problem with `'..' in path` check because it is now split by `os.sep == \`, not by `posixpath.sep == /`. Is it worth to revert this change? Also, do these tests also run on a linux box? It would probably be good to check if nothing has been broken there as well.

**Comment by serena-ruan** [comment]
> Well, judging by the test output, pathlib now works fine. All failed test cases under the `test_validate_path_is_safe_windows_good` are false positives and will be indeed treated by Python as absolute paths, so they should better be moved to the `test_validate_path_is_safe_windows_bad` list of paths. However, there is indeed a problem with `'..' in path` check because it is now split by `os.sep == \`, not by `posixpath.sep == /`. Is it worth to revert this change? Also, do these tests also run on a linux box? It would probably be good to check if nothing has been broken there as well.

Since we don't allow '\' in the path, maybe we just explicitly say `.split('/')`  instead of `.split(os.sep)`? This test is only run on windows because the same path might be disallowed on windows but allowed on linux right? `test_validate_path_is_safe_bad` only runs on linux, with ```
"/path",
        "../path",
        "../../path",
        "./../path",
        "path/../to/file",
        "path/../../to/file",
        "/etc/passwd",
        "/etc/passwd%00.jpg",
        "/etc/passwd%00.html",
        "/etc/passwd%00.txt",
        "/etc/passwd%00.php",
        "/etc/passwd%00.asp",
```

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
@adeadfed Thanks a lot for the help :)

**Comment by adeadfed** [comment]
Hey @harupy & @serena-ruan ! Was a pleasure to work with you to make the MLFlow more secure! I've just one question left to you: would you be willing to file a Github security advisory & validate the corresponsing reports on the huntr.dev platform for #8999 & #9053 now that they have been merged into the main branch? I would greatly appreciate that! Stay safe!

**Comment by serena-ruan** [comment]
> Hey @harupy & @serena-ruan ! Was a pleasure to work with you to make the MLFlow more secure! I've just one question left to you: would you be willing to file a Github security advisory & validate the corresponsing reports on the huntr.dev platform for #8999 & #9053 now that they have been merged into the main branch? I would greatly appreciate that! Stay safe!

Github security advisory sounds like a good plan! @harupy WDYT? I'll go to huntr.dev and mark your reported issues as resolved! Thanks so much 😃

**Pull request merged**